pymc_marketing/mmm/base.py

```diff
@@ -390,24 +399,198 @@ def plot_contribution_curves(self) -> plt.Figure:
+        ------
+        ValueError
+            If any of the required parameters are not provided or have an incorrect type.
+
+        Notes
+        -----
+        - The function optimally allocates the budget among different channels based on the provided estimations and budget constraints.
+        - It utilizes a cost function that calculates the contribution for each channel based on a quadratic curve fit to the data.
+        - The budget allocation is subject to a constraint to ensure the total allocated budget matches the total budget.
+        - The function uses optimization techniques to find the optimal budget allocation that maximizes contribution.
+        - The updated DataFrame includes an 'Optimal Budget' column with the allocated budget for each channel.
+        - The function also prints the maximum total contribution with the allocated budget and the proposed budget amount.
+        """
+
+        if total_budget is None or df_estimations is None:
+            raise ValueError(
+                "Total Budget and Estimations dataframe parameters must be provided."
+            )
+
+        if not isinstance(total_budget, (int, float)):
+            raise ValueError(
+                "The 'total_budget' parameter must be an integer or float."
+            )
+
+        if not isinstance(df_estimations, pd.DataFrame):
+            raise ValueError(
+                "The 'df_estimations' parameter must be a pandas DataFrame."
+            )
+
+        if not isinstance(budget_bounds, dict):
+            raise ValueError("The 'budget_bounds' parameter must be a dictionary.")
+
+        df = df_estimations.copy()
+        channel_contributions = self.compute_channel_contribution_original_scale().mean(
+            ["chain", "draw"]
+        )
+
+        lower_bounds = []
+        upper_bounds = []
+
+        for channel in df["channel"]:
+            if channel in budget_bounds:
+                min_bound, max_bound = budget_bounds[channel]
+            else:
+                min_bound = np.quantile(self.data[channel], 0.25)
+                max_bound = df.loc[df["channel"] == channel, "plateau_point"].values[0]
+
+            lower_bounds.append(min_bound)
+            upper_bounds.append(max_bound)
+
+        bounds = Bounds(lower_bounds, upper_bounds)
+        cons = {"type": "eq", "fun": budget_constraint, "args": (total_budget,)}
+
+        initial_guess = lower_bounds  # Start with minimum allocation for all channels
+        result = minimize(
+            cost_function,
+            initial_guess,
+            args=(df, self.data, channel_contributions),
+            bounds=bounds,
+            constraints=cons,
+        )
+
+        # Add the new column to the dataframe
+        df["Optimal Budget"] = np.round(result.x)
+
+        print("The maximum total contribution with these allocations is:")
+        print(-result.fun)
+
+        print("The proposed budget is:")
+        print(np.round(result.x.sum()))
```

Comment:
I am not super keen on having print statements ... shall we output a dataframe ?